src/Servers/Kestrel/Core/src/Internal/Http/ForChunkedEncodingMessageBody.cs

```diff
@@ -0,0 +1,670 @@
+
+            var trailerBuffer = buffer.Slice(0, 2);
+            var trailerSpan = trailerBuffer.ToSpan();
+
+            if (trailerSpan[0] == '\r' && trailerSpan[1] == '\n')
+            {
+                consumed = trailerBuffer.End;
+                examined = trailerBuffer.End;
+                AddAndCheckConsumedBytes(2);
+                _mode = Mode.Complete;
+            }
+            else
+            {
+                _mode = Mode.TrailerHeaders;
+            }
+        }
+
+        private int CalculateChunkSize(int extraHexDigit, int currentParsedSize)
+        {
+            try
+            {
+                checked
+                {
+                    if (extraHexDigit >= '0' && extraHexDigit <= '9')
+                    {
+                        return currentParsedSize * 0x10 + (extraHexDigit - '0');
+                    }
+                    else if (extraHexDigit >= 'A' && extraHexDigit <= 'F')
+                    {
+                        return currentParsedSize * 0x10 + (extraHexDigit - ('A' - 10));
+                    }
+                    else if (extraHexDigit >= 'a' && extraHexDigit <= 'f')
+                    {
+                        return currentParsedSize * 0x10 + (extraHexDigit - ('a' - 10));
+                    }
+                }
+            }
+            catch (OverflowException ex)
+            {
+                throw new IOException(CoreStrings.BadRequest_BadChunkSizeData, ex);
+            }
+
+            BadHttpRequestException.Throw(RequestRejectionReason.BadChunkSizeData);
+            return -1; // can't happen, but compiler complains
+        }
+        private ReadResult _readResult;
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            var dataLength = _readResult.Buffer.Slice(_readResult.Buffer.Start, consumed).Length;
+            _requestBodyPipe.Reader.AdvanceTo(consumed, examined);
+            OnDataRead(dataLength);
+        }
+
+        public override bool TryRead(out ReadResult readResult)
+        {
+            TryStart();
+
+            var res =_requestBodyPipe.Reader.TryRead(out _readResult);
+            readResult = _readResult;
+
+            if (_readResult.IsCompleted)
+            {
+                TryStop();
+            }
+            return res;
+        }
+
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            TryStart();
+
+            while (true)
+            {
+                _readResult = await StartTimingReadAsync(cancellationToken);
+                var readableBuffer = _readResult.Buffer;
+                var readableBufferLength = readableBuffer.Length;
+                StopTimingRead(readableBufferLength);
+
+                if (_readResult.IsCanceled)
```

Comment:
I don't believe I need the _userCanceled check here as there is still a RequestPipe. @halter73 you mentioned that the ConnectionPipe can be canceled at practically any point, but I don't believe it applies to the RequestPipe. I think I can remove the bool here and if IsCanceled, break out of the loop. 

Actually, I believe I can remove the while loop entirely if that's the case.